ws/context: Increase `write()` visibility to public

Turbo87 opened the pull request:
I'm running a TCP socket to WebSocket gateway using actix and I noticed that the messages that I relay to the WebSocket clients are copied once per client as part of `ws::Frame::message()`. I would like to avoid that overhead by calling `ws::Frame::message()` only once, and passing the result to each of the WebSocket clients.

This PR changes the `WebsocketContext::write()` method to be public, so that users are able to pass in `Binary` directly including the framing.

Turbo87 committed:
ws/context: Add docs to `write()` method

Turbo87 committed:
ws/context: Increase `write()` visibility to public

DoumanAsh commented on src/ws/context.rs:
```diff
@@ -133,7 +133,7 @@ where
 {
     /// Write payload
```
Please describe in documentation to this method how user should create `Binary` as user should not manually initialize it.
and instead use proper constructor.

Turbo87 replied:
Docs added 👍

DoumanAsh replied:
thanks!

DoumanAsh commented on src/ws/context.rs:
```diff
@@ -133,7 +133,7 @@ where
     /// Write payload
     #[inline]
-    fn write(&mut self, data: Binary) {
+    pub fn write(&mut self, data: Binary) {
```
Do you think it may be wise for us to introduce wrapper type for websocket frame, to avoid potential misuse?(e.g. user fill Binary himself)

Turbo87 replied:
I guess if `ws::Frame::message()` would return a dedicated type and `WebsocketContext::write()` would only accept that type it might be useful. I'm not sure though if `ws::Frame::message()` is used anywhere else though.

DoumanAsh replied:
I think if we want to expose this API, it should be dedicated type to avoid potential misuse, as you imagine not all people would be careful enough when reading documentation :)

You could leave this `write` method for internal use and instead create new method `send_frame` or just `frame` that would accept wrapper type and pass it's Binary data into `write`

Turbo87 replied:
I've introduced a `FramedBinary` type now for this purpose. This is technically a breaking change, but since master is using a newer `actix` version the next release would be a breaking release anyway 😉

fafhrd91 commented on src/ws/context.rs:
```diff
@@ -132,8 +132,13 @@ where
+    /// that handle the framing for you.
     #[inline]
-    fn write(&mut self, data: Binary) {
+    pub fn write(&mut self, data: Binary) {
```
maybe rename it to `write_raw()`?

Turbo87 replied:
good idea. renamed :)

Turbo87 committed:
ws/context: Rename `write()` to `write_raw()`

Turbo87 committed:
ws/frame: Introduce `FramedBinary` type

This type is introduced to avoid confusion between the `.binary()` and `.write_raw()` methods on WebSocket contexts

DoumanAsh commented on src/ws/frame.rs:
```diff
@@ -372,6 +374,10 @@ impl fmt::Display for Frame {
 
+/// A `Binary` representing a `WebSocket` message with framing.
+#[derive(Debug)]
+pub struct FramedBinary(pub Binary);
```
`pub` allows user to manually create the instance, I think we should prohibit it?

User should only correctly framed messages.

Turbo87 replied:
If I make it private or `pub(crate)` it complains about "can't leak private type" and won't compile

DoumanAsh replied:
Oh what I mean is that its member `pub Binary` should be just `Binary` so that user couldn't instantiate it himself like `FramedBianry(Bytes::new().into())`

Turbo87 replied:
actually since the `frame` module is not `pub` and the struct is not exported it already is private and only accessible inside of `actix-web` so there is no change needed here

DoumanAsh replied:
In that case we should re-export it as public.
This `FramedBinary` will be leaked through `Frame::message` while it does prevents user fron instantiating it, it also prevents user from reading any documentation on it (you'll have blanked type `FramedBinary` without any information about it)
I think it would not be very nice for reader?

Turbo87 replied:
done

DoumanAsh commented on src/ws/frame.rs:
```diff
@@ -372,6 +374,10 @@ impl fmt::Display for Frame {
     }
 }
 
+/// A `Binary` representing a `WebSocket` message with framing.
```
I think it would be better to say just:

>`WebSocket` framed message

We don't really need to overload users :)

Turbo87 replied:
updated

DoumanAsh commented on src/ws/frame.rs:
```diff
@@ -372,6 +374,10 @@ impl fmt::Display for Frame {
 
+/// A `Binary` representing a `WebSocket` message with framing.
+#[derive(Debug)]
+pub struct FramedBinary(pub Binary);
```
Maybe name it `FramedMessage`?

Even `write_raw` says now 😄 

>This is a low-level function that accepts framed messages

Turbo87 replied:
done

DoumanAsh reviewed (commented):
I still have concern about `FramedBinary` being instantiated by user

Turbo87 committed:
ws/frame: Rename `FramedBinary` type to `FramedMessage`

Turbo87 committed:
ws/frame: Adjust `FramedMessage` doc comment

Turbo87 committed:
ws: Export `FramedMessage` struct

DoumanAsh commented on src/ws/frame.rs:
```diff
@@ -372,6 +374,10 @@ impl fmt::Display for Frame {
 
+/// `WebSocket` message with framing.
+#[derive(Debug)]
+pub struct FramedMessage(pub Binary);
```
`pub Binary` -> `Binary` 😄

Turbo87 replied:
```
error[E0611]: field `0` of tuple-struct `ws::frame::FramedMessage` is private
   --> src/ws/client.rs:535:45
    |
535 |             let _ = inner.tx.unbounded_send(data.0.take());
    |                                             ^^^^^^

error[E0611]: field `0` of tuple-struct `ws::frame::FramedMessage` is private
   --> src/ws/context.rs:147:50
    |
147 |             stream.push(ContextFrame::Chunk(Some(data.0)));
    |                                                  ^^^^^^
```

Turbo87 replied:
I could make it `pub(crate)` though if that helps?

DoumanAsh replied:
Oh I see, sorry, I didn't notice it is used outside of this module.
Yes, it should be public only within crate itself, we cannot let user to write non-framed data.

Turbo87 replied:
done

Turbo87 committed:
ws/frame: Change `FramedMessage` content visibility

DoumanAsh commented:
@Turbo87 Thanks
